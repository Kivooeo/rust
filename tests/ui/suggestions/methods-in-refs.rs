fn main() {
    let a = A;
    a.new();
    //~^ ERROR no method named `new` found for struct `A` in the current scope [E0599]
    let mut b = B;
    b.new();
    //~^ ERROR no method named `new` found for struct `B` in the current scope [E0599]
    let c = C;
    c.new();
    //~^ ERROR no method named `new` found for struct `C` in the current scope [E0599]
    let d = D;
    d.new();
    //~^ ERROR no method named `new` found for struct `D` in the current scope [E0599]
}
// Cases where methods was successfully found in any of &/& mut implementations
struct A;

trait X {
    fn new(&self) -> i32;
}

impl X for &A {
    fn new(&self) -> i32 {
        0
    }
}

struct B;

trait Y {
    fn new(&mut self) -> i32;
}

impl Y for &mut B {
    fn new(&mut self) -> i32 {
        0
    }
}

// Cases where methods was not successfully found in any of &/& mut implementations
struct C;

struct D;

trait Z {
    fn new(&self);
}
