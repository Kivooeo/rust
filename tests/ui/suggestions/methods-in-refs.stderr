error[E0599]: no method named `new` found for struct `A` in the current scope
  --> $DIR/methods-in-refs.rs:3:7
   |
LL |     a.new();
   |       ^^^ method not found in `A`
...
LL | struct A;
   | -------- method `new` not found for this struct
   |
help: `new` method was found in `&A`, you can access it with `.as_ref()`
   |
LL |     a.as_ref().new();
   |       +++++++++

error[E0599]: no method named `new` found for struct `B` in the current scope
  --> $DIR/methods-in-refs.rs:6:7
   |
LL |     b.new();
   |       ^^^ method not found in `B`
...
LL | struct B;
   | -------- method `new` not found for this struct
   |
help: `new` method was found in `&mut B`, you can access it with `.as_mut()`
   |
LL |     b.as_mut().new();
   |       +++++++++

error[E0599]: no method named `new` found for struct `C` in the current scope
  --> $DIR/methods-in-refs.rs:9:7
   |
LL |     c.new();
   |       ^^^ method not found in `C`
...
LL | struct C;
   | -------- method `new` not found for this struct
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `new`, perhaps you need to implement one of them:
           candidate #1: `X`
           candidate #2: `Y`
           candidate #3: `Z`

error[E0599]: no method named `new` found for struct `D` in the current scope
  --> $DIR/methods-in-refs.rs:12:7
   |
LL |     d.new();
   |       ^^^ method not found in `D`
...
LL | struct D;
   | -------- method `new` not found for this struct
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `new`, perhaps you need to implement one of them:
           candidate #1: `X`
           candidate #2: `Y`
           candidate #3: `Z`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
